packages/react-essentials/hooks/useUpdateLoop.test.ts

```diff
@@ -0,0 +1,269 @@
+import { act, renderHook } from "@testing-library/react";
+import { beforeEach, afterEach, describe, expect, it, vi } from "vitest";
+import { useUpdateLoop } from "./useUpdateLoop";
+
+beforeEach(() => {
+  vi.useFakeTimers();
+
+  vi.spyOn(window, "requestAnimationFrame").mockImplementation(
+    (callback: FrameRequestCallback): number => {
+      /**
+       * A timeout of `1000 / 50` is used to simulate a display with a 50Hz
+       * refresh rate. This value was chosen because it results in a timeout
+       * that is an integer value, which makes the number of frames easier to
+       * calculate.
+       */
+      setTimeout(callback, 1000 / 50);
+      return Math.random();
+    },
+  );
+
+  vi.spyOn(window, "cancelAnimationFrame").mockImplementation(
+    (id: number): void => {
+      clearTimeout(id);
+      vi.clearAllTimers();
+    },
+  );
+});
+
+afterEach(() => {
+  vi.runOnlyPendingTimers();
+  vi.useRealTimers();
+  vi.restoreAllMocks();
+});
+
+describe("useUpdateLoop", () => {
+  it("should call the onUpdate callback on each frame", () => {
+    const onUpdate = vi.fn();
+
+    renderHook(() => useUpdateLoop({ onUpdate, immediately: true }));
+
+    vi.advanceTimersByTime(1000);
+
+    expect(onUpdate).toHaveBeenCalledTimes(50);
+  });
+
+  it("should support starting the loop manually", () => {
+    const onUpdate = vi.fn();
+
+    const { result } = renderHook(() =>
+      useUpdateLoop({
+        onUpdate,
+        immediately: false,
+      }),
+    );
+
+    vi.advanceTimersByTime(500);
+
+    expect(onUpdate).toHaveBeenCalledTimes(0);
+
+    act(() => {
+      result.current.start();
+    });
+
+    vi.advanceTimersByTime(500);
+
+    expect(onUpdate).toHaveBeenCalledTimes(25);
+  });
+
+  it("should call the onUpdate callback with the correct frame data", () => {
+    const onUpdate = vi.fn();
+
+    renderHook(() => useUpdateLoop({ onUpdate, immediately: true }));
+
+    vi.advanceTimersByTime(1000);
```

Comment:
Would it be an idea to do this several times? At least two. Maybe even different timing `vi.advanceTimersByTime(5000);`